### PR TITLE
User namespaces: graduate from experimental

### PR DESCRIPTION
**Pull request by estesp**
Moves experimental user namespaces to master in preparation for the 1.10 release. Adds documentation to regular `docker daemon` command docs.  Properly warns of incompatible flags for namespace sharing + privileged.

Closes: #18448
Closes: #18430
Closes: #18247

### CODE REVIEW

**Comment by estesp** [comment]
ping @thaJeztah : I spent some extra time trying to make the docs more readable/clear.  Feel free to start reviewing/providing feedback as I think the docs may need more help than the code, which is basically just a migration of the flag and settings.

**Comment by thaJeztah** [comment]
ping @moxiegirl as well :smile:

**Comment by vdemeester** [comment]
:heart_eyes_cat:

**Comment by jfrazelle** [comment]
yayayayayayayayayayayayyayaaya! LGTM

**Comment by LK4D4** [comment]
yolo
LGTM

**Comment by icecrime** [comment]
Looks YOLO To Me :+1:

**Comment by estesp** [comment]
@jfrazelle just FYI--updated the patch to stop `hack/make.sh` from adding `experimental` build tag when `DOCKER_REMAP_ROOT` is set, given it will now be "testable" from master.  Let me know if you see any issues with that, but the CI **userns** run should verify that's correct.

**Comment by jfrazelle** [comment]
sweet thats perfect thanks!

**Comment by thaJeztah** [comment]
@estesp don't forget to update the daemon man page as well, to include the new flag; https://github.com/docker/docker/blob/master/man/docker-daemon.8.md

**Comment by HackToday** [comment]
@thaJeztah  do you know how to test it when it was not merged ? Do we have binary build for that ? Or need I cherry-pick these code in my repo and build binary by myself ?  Thanks

**Comment by vdemeester** [comment]
@HackToday You can try out a pull request by [checking it out using git](https://help.github.com/articles/checking-out-pull-requests-locally/) and then build it (using `make binary` for example :wink:)

**Comment by HackToday** [comment]
Thanks @vdemeester  I tried this clone 
https://github.com/estesp/docker/tree/lets-do-this

Since I did not figure it out just cherry-pick easy ways to do that :(

**Comment by HackToday** [comment]
Found following potential issues:

1. use host with --uts=host or --ipc=host failed with following strange error message:


root@003d1da3734a:/go/src/github.com/docker/docker# docker run -it --uts=host   myhtop  pidtest-1
mount: permission denied
Could not mount /sys/kernel/security.
AppArmor detection and --privileged mode might break.
mount: permission denied
root@003d1da3734a:/go/src/github.com/docker/docker# docker run -it --ipc=host   myhtop  pidtest-1
mount: permission denied
Could not mount /sys/kernel/security.
AppArmor detection and --privileged mode might break.
mount: permission denied


2.  for --ipc use container:ID

root@003d1da3734a:/go/src/github.com/docker/docker# docker ps
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
5040226021d9        busybox             "sh"                23 minutes ago      Up 23 minutes                           testok
2dc6bb7c5ebb        busybox             "sh"                25 minutes ago      Up 25 minutes                           goofy_northcutt
root@003d1da3734a:/go/src/github.com/docker/docker# 
root@003d1da3734a:/go/src/github.com/docker/docker# 
root@003d1da3734a:/go/src/github.com/docker/docker# docker run --name testsecond  -it --ipc=container:testok  busybox sh
docker: Error response from daemon: Container command could not be invoked..
root@003d1da3734a:/go/src/github.com/docker/docker# docker ps -a
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
98268acac542        busybox             "sh"                8 seconds ago       Created                                 testsecond
5040226021d9        busybox             "sh"                23 minutes ago      Up 23 minutes                           testok
2dc6bb7c5ebb        busybox             "sh"                25 minutes ago      Up 25 minutes                           goofy_northcutt


And the container was created.

**Comment by HackToday** [comment]
And another issue:

When run with docker run --read-only -v /icanwrite busybox sh  still have long go trace message like below:

--
5: main
Package: main
File: docker.go@18
---
6: main
Package: runtime
File: proc.go@111
---
7: goexit
Package: runtime
File: asm_amd64.s@1721
docker: Error response from daemon: Cannot start container c8bd26e061fe5d4b229581c912b0204c0dee4b3bebf82e72940895591bf74f09: [9] System error: operation not permitted.

It seems still an issue mentioned like this one https://github.com/docker/docker/issues/18430 (although the old one is about pid namespace)

File: daemon/daemon_unix.go [hunk 1]
```diff
@@ -375,6 +380,18 @@ func verifyPlatformContainerSettings(daemon *Daemon, hostConfig *containertypes.
...
+	// check for various conflicting options with user namespaces
+	if daemon.configStore.RemappedRoot != "" {
+		if hostConfig.Privileged {
+			return warnings, fmt.Errorf("Privileged mode is incompatible with user namespaces.")
+		}
+		if hostConfig.NetworkMode.IsHost() || hostConfig.NetworkMode.IsContainer() {
+			return warnings, fmt.Errorf("Cannot share network namespaces with user namespaces enabled.")
+		}
+		if hostConfig.PidMode.IsHost() {
+			return warnings, fmt.Errorf("Cannot use the host PID namespace with user namespaces enabled.")
```
Comment (HackToday) [inline daemon/daemon_unix.go]: Suggest replace use with share, as keep consistent above network namespace error message.

File: daemon/daemon_unix.go [hunk 2]
```diff
@@ -375,6 +380,18 @@ func verifyPlatformContainerSettings(daemon *Daemon, hostConfig *containertypes.
 		warnings = append(warnings, "IPv4 forwarding is disabled. Networking will not work.")
 		logrus.Warnf("IPv4 forwarding is disabled. Networking will not work")
 	}
+	// check for various conflicting options with user namespaces
+	if daemon.configStore.RemappedRoot != "" {
+		if hostConfig.Privileged {
+			return warnings, fmt.Errorf("Privileged mode is incompatible with user namespaces.")
+		}
+		if hostConfig.NetworkMode.IsHost() || hostConfig.NetworkMode.IsContainer() {
+			return warnings, fmt.Errorf("Cannot share network namespaces with user namespaces enabled.")
```
Comment (HackToday) [inline daemon/daemon_unix.go]: also sometime I found we use "namespaces"(like user and network), sometime we use "namespace" (like pid)

Could you tell me What's the difference ?  Thanks

Comment (estesp) [inline daemon/daemon_unix.go]: There is not any specific difference other than what sounds "right" in English when talking about a singular versus plural concept.  I have updated the messages to be more uniform--please let me know if it makes sense to you.

File: daemon/daemon_unix.go (hunk 1 above)
Comment (estesp) [inline daemon/daemon_unix.go]: Done.

**Comment by estesp** [comment]
> And another issue:
>
> When run with docker run --read-only -v /icanwrite busybox sh still have long go trace message like below:

I have added an early error similar to the namespace sharing errors when using `--read-only=true`.  The docs mention this is not possible with user namespaces, but adding the error makes it much clearer to the user.  Thanks.

**Comment by estesp** [comment]
@thaJeztah I have made the modifications you suggested as well as added the additional info to the `docker-daemon` man page.  Please let me know if the man page additions are sufficient of if the more complete detail from the markdown docs should be added here as well.

@HackToday I have made corrections to the docs per your suggestions and added an error message when `--read-only` is in use (to prevent the long complicated Go dump output).  Your other issues (with shared ipc/uts) seem to be related to something your specific container is doing at startup?  If I run a simple "busybox" or "ubuntu" image, I don't get these errors.  I can look into it further, but maybe you can try with a standard image and then understand what your "**myhtop**" container is trying to do

**Comment by HackToday** [comment]
@estesp I will pick up your changes and do a test, will update results today. Thanks

File: daemon/daemon_unix.go [hunk 3]
```diff
@@ -375,6 +380,21 @@ func verifyPlatformContainerSettings(daemon *Daemon, hostConfig *containertypes.
...
+			return warnings, fmt.Errorf("Privileged mode is incompatible with user namespaces.")
+		}
+		if hostConfig.NetworkMode.IsHost() || hostConfig.NetworkMode.IsContainer() {
+			return warnings, fmt.Errorf("Cannot share the host or a container's network namespace when user namespaces are enabled.")
+		}
+		if hostConfig.PidMode.IsHost() {
+			return warnings, fmt.Errorf("Cannot share the host PID namespace when user namespaces are enabled.")
+		}
+		if hostConfig.ReadonlyRootfs {
+			return warnings, fmt.Errorf("Cannot use the --read-only option when user namespaces enabled.")
```
Comment (jamtur01) [inline daemon/daemon_unix.go]: when user namespaces are enabled.

File: docs/reference/commandline/daemon.md [hunk 4]
```diff
@@ -628,6 +629,131 @@ For information about how to create an authorization plugin, see [authorization
 plugin](../../extend/authorization.md) section in the Docker extend section of this documentation.
 
 
+## Daemon user namespace options
+
+Linux kernel [user namespace support](http://man7.org/linux/man-pages/man7/user_namespaces.7.html) provides additional security by enabling
```
Comment (jamtur01) [inline docs/reference/commandline/daemon.md]: The Linux kernel....

File: docs/reference/commandline/daemon.md [hunk 5]
```diff
@@ -628,6 +629,131 @@ For information about how to create an authorization plugin, see [authorization
...
+a process--and therefore a container--to have a unique range of user and
```
Comment (jamtur01) [inline docs/reference/commandline/daemon.md]: single dashes if you must use dashes.

File: docs/reference/commandline/daemon.md [hunk 6]
```diff
@@ -628,6 +629,131 @@ For information about how to create an authorization plugin, see [authorization
...
+`dockremap`, and entries will be created for it in `/etc/passwd` and
+`/etc/group` using your distro's standard user and group creation tools.
+
+> **Note**: The single mapping per-daemon restriction is in place for now
+> because Docker shares image layers from its local cache across all
+> containers running on the engine instance.  Since file ownership must be
+> the same for all containers sharing the same layer content, the decision
+> was made to map the file ownership on `docker pull` to the daemon's user and
+> group mappings so that there is no delay for running containers once the
+> content is downloaded--exactly the same performance characteristics as with
```
Comment (jamtur01) [inline docs/reference/commandline/daemon.md]: single dash.

File: docs/reference/commandline/daemon.md [hunk 7]
```diff
@@ -628,6 +629,131 @@ For information about how to create an authorization plugin, see [authorization
...
+these resources are name-based, not id-based.  If the numeric ID information
+provided does not exist as entries in `/etc/passwd` or `/etc/group`, daemon
+startup will fail with an error message.
+
+*Example: starting with default Docker user management:*
+
+```
+     $ docker daemon --userns-remap=default
+```    
+When `default` is provided, Docker will create--or find the existing--user and group
```
Comment (jamtur01) [inline docs/reference/commandline/daemon.md]: Single dashes.

File: docs/reference/commandline/daemon.md [hunk 8]
```diff
@@ -628,6 +629,131 @@ For information about how to create an authorization plugin, see [authorization
...
+The following standard Docker features are currently incompatible when
+running a Docker daemon with user namespaces enabled:
+
+ - sharing PID or NET namespaces with the host (`--pid=host` or `--net=host`)
+ - sharing a network namespace with an existing container (--net=container:*other*)
+ - A `--readonly` container filesystem (this is a Linux kernel restriction against remounting with modified flags of a currently mounted filesystem when inside a user namespace)
+ - external (volume or graph) drivers which are unaware/incapable of using daemon user mappings
+ - Using `--privileged` mode flag on `docker run`
+
+Note that in general, user namespaces are an advanced feature and will require
```
Comment (jamtur01) [inline docs/reference/commandline/daemon.md]: If this is a note it should be formatted as one.

File: docs/reference/commandline/daemon.md [hunk 9]
```diff
@@ -628,6 +629,131 @@ For information about how to create an authorization plugin, see [authorization
...
+coordination with other capabilities. For example, if volumes are mounted from
+the host, file ownership will have to be pre-arranged if the user or
+administrator wishes the containers to have expected access to the volume
+contents.
+
+Finally, while the `root` user inside a user namespaced container process has
+many of the expected admin privileges that go in hand with being **root**, the
```
Comment (jamtur01) [inline docs/reference/commandline/daemon.md]: Perhaps?

    being the \`root\` user ...

File: docs/reference/commandline/daemon.md [hunk 10]
```diff
@@ -628,6 +629,131 @@ For information about how to create an authorization plugin, see [authorization
...
+Linux kernel has restrictions based on its knowledge that this is a user namespaced
+process. The most notable restriction that we are aware of at this time is the
+inability to use `mknod`. Permission will be denied for device creation even as
+container root inside a user namespace.
```
Comment (jamtur01) [inline docs/reference/commandline/daemon.md]: container \`root\` ???

**Comment by jamtur01** [comment]
Nice Docs - minor edits.

**Comment by estesp** [comment]
Thanks for the docs review @jamtur01. I have updated the PR per your comments.

**Comment by jamtur01** [comment]
One minor fix but otherwise Docs LGTM. Thanks!

**Comment by estesp** [comment]
Great! @jamtur01 I've used "superuser" instead, and the PR is updated.  Thanks!

**Comment by cyphar** [comment]
I think that #19251 should be considered before we make this non-experimental (quite a few systems don't have the latest version of shadow).

**Comment by HackToday** [comment]
LGTM

**Comment by cyphar** [comment]
@HackToday "LGTM"s should only be used by Docker maintainers (they are used by other maintainers to figure out whether there is a consensus on a feature being merged). It's just confusing to have non-maintainers LGTM things.

**Comment by HackToday** [comment]
@cyphar  sure, Ignore LGTM,   I will +1 about this

**Comment by HackToday** [comment]
:+1:

**Comment by cpuguy83** [comment]
ping @thaJeztah for docs signoff

**Comment by thaJeztah** [comment]
docs LGTM, let's merge!

**Comment by tiborvass** [comment]
\o/

**Pull request merged**